readline/buffer.go

```diff
@@ -121,114 +163,258 @@ func (b *Buffer) MoveToEnd() {
 			remainder := b.Size() % b.LineWidth
 			fmt.Printf(CursorBOL + cursorRightN(len(b.Prompt.prompt())+remainder))
 		} else {
-			fmt.Print(cursorRightN(b.Size() - b.Pos))
+			fmt.Print(cursorRightN(b.Size() - b.DisplayPos))
 		}
 
-		b.Pos = b.Size()
+		b.Pos = b.Buf.Size()
+		b.DisplayPos = b.Size()
 	}
 }
 
 func (b *Buffer) Size() int {
-	return b.Buf.Size()
+	sum := 0
+	for i := 0; i < b.Buf.Size(); i++ {
+		if e, ok := b.Buf.Get(i); ok {
+			if r, ok := e.(rune); ok {
+				sum += runewidth.RuneWidth(r)
+			}
+		}
+	}
+
+	return sum
 }
 
 func (b *Buffer) Add(r rune) {
+	rLength := runewidth.RuneWidth(r)
+
 	if b.Pos == b.Buf.Size() {
```

Comment:
You can DRY this part out w/ the part below by making a new method called something like
```
func (b *Buffer) addChar(r rune, insert bool) {
    ...
}
```

then:
```
    if b.Pos == b.Buf.Size() {
        b.addChar(r, false)
    } else {
        b.addChar(r, true)
    }
```